src/v/security/tests/plain_authenticator_test.cc

```diff
@@ -0,0 +1,242 @@
+        ASSERT_FALSE_CORO(res.has_value()) << "Should not have authenticated";
+        EXPECT_EQ(res.assume_error(), security::errc::invalid_credentials);
+        EXPECT_FALSE(authn->complete());
+        EXPECT_TRUE(authn->failed());
+    }
+}
+
+TEST_F_CORO(sasl_plain_test_fixture, test_authz_id) {
+    ss::sstring username("user");
+    ss::sstring password("password");
+    auto authn_msg = bytes::from_string("authz");
+    auto authn_msg_tmp = create_authn_message(username, password);
+    std::copy(
+      std::make_move_iterator(authn_msg_tmp.begin()),
+      std::make_move_iterator(authn_msg_tmp.end()),
+      std::back_inserter(authn_msg));
+
+    auto res = co_await authn->authenticate(authn_msg);
+    ASSERT_FALSE_CORO(res.has_value()) << "Should not have authenticated";
+    EXPECT_EQ(res.assume_error(), security::errc::invalid_credentials);
+    EXPECT_FALSE(authn->complete());
+    EXPECT_TRUE(authn->failed());
+}
+
+TEST_F_CORO(sasl_plain_test_fixture, no_username) {
+    ss::sstring password("password");
+    auto res = co_await authn->authenticate(create_authn_message("", password));
+    ASSERT_FALSE_CORO(res.has_value()) << "Should not have authenticated";
+    EXPECT_EQ(res.assume_error(), security::errc::invalid_credentials);
+    EXPECT_FALSE(authn->complete());
+    EXPECT_TRUE(authn->failed());
+}
+
+TEST_F_CORO(sasl_plain_test_fixture, no_password) {
+    ss::sstring username("user");
+    auto res = co_await authn->authenticate(create_authn_message(username, ""));
+    ASSERT_FALSE_CORO(res.has_value()) << "Should not have authenticated";
+    EXPECT_EQ(res.assume_error(), security::errc::invalid_credentials);
+    EXPECT_FALSE(authn->complete());
+    EXPECT_TRUE(authn->failed());
+}
+
+TEST_F_CORO(sasl_plain_test_fixture, massive_username) {
+    ss::sstring username(1024, 'a');
+    ss::sstring password("password");
+
+    auto creds = make_scram_credential(
+      scram_type_t::scram_sha256, username, password);
+    store.put(security::credential_user{username}, creds);
+
+    auto res = co_await authn->authenticate(
+      create_authn_message(username, password));
+    ASSERT_FALSE_CORO(res.has_value()) << "Should not have authenticated";
+    EXPECT_EQ(res.assume_error(), security::errc::invalid_credentials);
+    EXPECT_FALSE(authn->complete());
+    EXPECT_TRUE(authn->failed());
+}
+
+TEST_F_CORO(sasl_plain_test_fixture, massive_password) {
+    ss::sstring username("user");
+    ss::sstring password(1024, 'a');
+
+    auto creds = make_scram_credential(
+      scram_type_t::scram_sha256, username, password);
+    store.put(security::credential_user{username}, creds);
+
+    auto res = co_await authn->authenticate(
+      create_authn_message(username, password));
+    ASSERT_FALSE_CORO(res.has_value()) << "Should not have authenticated";
+    EXPECT_EQ(res.assume_error(), security::errc::invalid_credentials);
+    EXPECT_FALSE(authn->complete());
+    EXPECT_TRUE(authn->failed());
+}
```

Comment:
Should we add a test that makes some assertions on the audit user as well?

Reply:
Just a simple test that verifies that the audit user's name is properly set on success. If you think it's too granular, I'm fine with that, but it would be a sanity check to ensure that the audit user is updated.